pkg/server/cloud/aws/network.go

```diff
@@ -316,57 +316,85 @@ func (e *AwsEC2) ModifySourceDestinationCheck(instanceID string, isEnabled bool)
 	return err
 }
 
-func (e *AwsEC2) RemoveRoute(destinationCIDR string) error {
+func (e *AwsEC2) RemoveRoute(destinationCIDR, instanceID string) error {
+	if destinationCIDR == "" || instanceID == "" {
+		return fmt.Errorf(
+			"invalid input: at least one non-empty value needed (got %q %q)",
+			destinationCIDR, instanceID)
+	}
 	out, err := e.client.DescribeRouteTables(&ec2.DescribeRouteTablesInput{
 		Filters: []*ec2.Filter{
 			{
-				Name: aws.String("vpc-id"),
-				Values: []*string{
-					aws.String(e.vpcID),
-				},
+				Name:   aws.String("vpc-id"),
+				Values: aws.StringSlice([]string{e.vpcID}),
+			},
+			{
+				Name:   aws.String("association.main"),
+				Values: aws.StringSlice([]string{"false"}),
+			},
+			{
+				Name:   aws.String("association.subnet-id"),
+				Values: aws.StringSlice([]string{e.subnetID}),
 			},
 		},
 	})
 	if err != nil {
-		return util.WrapError(err, "Could not list vpc route tables")
+		return util.WrapError(err, "listing vpc route tables")
 	}
 	for _, table := range out.RouteTables {
 		if len(table.Associations) == 0 {
 			continue
 		}
 		for _, route := range table.Routes {
-			if destinationCIDR == aws.StringValue(route.DestinationCidrBlock) {
-				_, err = e.client.DeleteRoute(&ec2.DeleteRouteInput{
-					DestinationCidrBlock: route.DestinationCidrBlock,
-					RouteTableId:         table.RouteTableId,
-				})
-				if err != nil {
-					return util.WrapError(
-						err,
-						"Error deleting old route table entry for table %s",
-						*table.RouteTableId,
-					)
-				}
-				break
+			if destinationCIDR != "" &&
+				destinationCIDR != aws.StringValue(route.DestinationCidrBlock) {
+				continue
+			}
+			// Ignore any routes that are not instance routes.
+			routeIID := aws.StringValue(route.InstanceId)
+			if routeIID == "" || instanceID != "" && instanceID != routeIID {
```

Comment:
This is kinda confusing.  I don't think you need the `instanceID != "" && ` part of this expression.  Just `routeIID == "" || instanceID != routeIID`.  Can you double check this? I might be wrong...

1. We've checked that instanceID != "" at the top of the function
2. If `instanceID == ""` then the check `instanceID != routeID` will evaluate to false only if `routeID == ""`.  That's already checked.

Reply:
I removed the check at the top of the function, since we are now cleaning up dangling instance routes via leaving both input variables empty.